src/FrameMapper.cpp

```diff
@@ -116,6 +117,15 @@ void FrameMapper::Init()
 	// Clear cache
 	final_cache.Clear();
 
+	// Get clip position from parent clip (if any)
+	float clipPosition = 0.0;
+	float clipStart = 0.0;
+	Clip *parent = (Clip *) ParentClip();
+	if (parent) {
+		clipPosition = parent->Position();
+		clipStart = parent->Start();
+	}
+
```

Comment:
Near as I can tell, this block of code (which is duplicated enough to send Codacy into conniptions) is only ever used to collect numbers to feed into `AdjustFrameNumber()`, yes?

So, why not just make **`AdjustFrameNumber()`** do the work of checking those values, by passing it the pointer?

```cpp
// In FrameMapper::Init() and friends
int remaining_samples = Frame::GetSamplesPerFrame(
    AdjustFrameNumber(frame_number, ParentClip()),
    target,
    reader->info.sample_rate,
    reader->info.channels
);

// Declared in FrameMapper.h
int64_t FrameMapper::AdjustFrameNumber(int64_t frame_number, ClipBase* parent);

// Defined in FrameMapper.cpp
int64_t FrameMapper::AdjustFrameNumber(int64_t frame_number, ClipBase* parent) {
    float position = parent ? parent->Position() : 0.0;
    float start = parent > parent->Start() : 0.0;

    // ...rest of code...
```
That way:
1. If `ParentClip()` returns `nullptr` that's fine, because it'll just get passed in to `AdjustFrameNumber` and you'll check it there.
1. You don't need any of this endlessly-duplicated setup code in the caller.

...Though — come to think of it, if `ParentClip()` is a method of the _FrameMapper_ instance, seems like you might not even need to pass **that** into `AdjustFrameNumber()`. Couldn't _it_ call `ParentClip()`?

Reply:
Yes, ParentClip() could be directly called in AdjustFrameNumber. I think my idea was to prevent calling ParentClip() over and over again... but the cost is actually trivial... so I like this idea.

Reply:
I've moved this duplicated logic into AdjustFrameNumber. Thanks.